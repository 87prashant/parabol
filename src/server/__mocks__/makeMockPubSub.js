const makeMockPubSub = () => ({
  publish: jest.fn(),
  subscribe: jest.fn()
});

export default makeMockPubSub;
