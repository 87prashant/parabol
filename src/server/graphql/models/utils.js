import {GraphQLNonNull, GraphQLInputObjectType} from 'graphql';
import getRethink from 'server/database/rethinkDriver';
import mime from 'mime-types';
import {
  APP_CDN_USER_ASSET_SUBDIR,
  APP_MAX_AVATAR_FILE_SIZE
} from 'universal/utils/constants';
import protocolRelativeUrl from 'server/utils/protocolRelativeUrl';
import {s3SignPutObject} from 'server/utils/s3';
import path from 'path';

// Stringify an object to handle multiple errors
// Wrap it in a new Error type to avoid sending it twice via the originalError field
export const errorObj = obj => new Error(JSON.stringify(obj));

export const handleSchemaErrors = (errors) => {
  if (Object.keys(errors).length > 0) {
    throw errorObj(errors);
  }
};

// VERY important, otherwise eg a user could "create" a new team with an existing teamId & force join that team
export const ensureUniqueId = async (table, id) => {
  const r = getRethink();
  const res = await r.table(table).get(id);
  if (res) {
    throw errorObj({type: 'unique id collision'});
  }
};

export const ensureUserInOrg = async (userId, orgId) => {
  const r = getRethink();
  const inOrg = await r.table('User').get(userId)('orgs').contains(orgId);
  if (!inOrg) {
    throw errorObj({type: `user ${userId} does not belong to org ${orgId}`});
  }
  return true;
};

// if the add & update schemas have different required fields, use this
export const nonnullifyInputThunk = (name, inputThunk, requiredFieldNames) => {
  return new GraphQLInputObjectType({
    name,
    fields: () => {
      const newFields = inputThunk();
      requiredFieldNames.forEach(fieldName => {
        newFields[fieldName].type = new GraphQLNonNull(newFields[fieldName].type);
      });
      return newFields;
    }
  });
};

function getFields(context, astsParams = context.fieldNodes) {
  // for recursion...Fragments doesn't have many sets...
  const asts = Array.isArray(astsParams) ? astsParams : [astsParams];

  // get all selectionSets
  const selectionSets = asts.reduce((selections, source) => {
    selections.push(...source.selectionSet.selections);
    return selections;
  }, []);

  // return fields
  return selectionSets.reduce((list, ast) => {
    switch (ast.kind) {
      case 'Field' :
        // eslint-disable-next-line no-param-reassign
        list[ast.name.value] = true;
        return list;
      case 'InlineFragment':
        return {
          ...list,
          ...getFields(context, ast)
        };
      case 'FragmentSpread':
        return {
          ...list,
          ...getFields(context, context.fragments[ast.name.value])
        };
      default:
        throw new Error('Unsuported query selection');
    }
  }, {});
}

export function getRequestedFields(refs) {
  const fieldsObj = getFields(refs);
  return Object.keys(fieldsObj);
}

export function firstChange(possiblyUpdatedResult) {
  if (possiblyUpdatedResult.changes.length) {
    if (possiblyUpdatedResult.changes.length > 1) {
      console.warn('firstChange() detects more than 1 change, returning 1st.');
    }
    return possiblyUpdatedResult.changes[0];
  }
  return { new_val: undefined, old_val: undefined };
}

export function updatedOrOriginal(possiblyUpdatedResult, original) {
  /*
   * There will only be changes to return if there were changes made to the
   * DB. Therefore, we've got to check.
   */
  return firstChange(possiblyUpdatedResult).new_val || original;
}

export function getOldVal(possiblyUpdatedResult) {
  return firstChange(possiblyUpdatedResult).old_val;
}

export function getNewVal(resultWithReturnChanges) {
  return firstChange(resultWithReturnChanges).new_val;
}

export function makeEnumValues(constArr) {
  return constArr.reduce((values, constant) => {
    values[constant] = {value: constant};
    return values;
  }, {})
}

export function validateAvatarUpload(contentType, contentLength) {
  if (typeof process.env.CDN_BASE_URL === 'undefined') {
    throw errorObj({_error: 'CDN_BASE_URL environment variable is not defined'});
  }
  if (!contentType || !contentType.startsWith('image/')) {
    throw errorObj({_error: 'file must be an image'});
  }
  const ext = mime.extension(contentType);
  if (!ext) {
    throw errorObj({_error: `unable to determine extension for ${contentType}`});
  }
  if (contentLength > APP_MAX_AVATAR_FILE_SIZE) {
    throw errorObj({_error: 'avatar image is too large'});
  }
  return ext;
}

export function getS3PutUrl(contentType, contentLength, partialPath) {
  const parsedUrl = protocolRelativeUrl.parse(process.env.CDN_BASE_URL);
  const pathname = path.join(parsedUrl.pathname,
    APP_CDN_USER_ASSET_SUBDIR,
    partialPath
  );
  return s3SignPutObject(
    pathname,
    contentType,
    contentLength,
    'public-read'
  );
}
