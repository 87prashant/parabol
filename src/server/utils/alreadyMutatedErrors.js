import sendAuthRaven from 'server/utils/sendAuthRaven';

export const sendAlreadyJoinedIntegrationError = (authToken, integrationId) => {
  const breadcrumb = {
    message: 'You are already a part of this integration',
    category: 'Already mutated',
    data: {integrationId}
  };
  return sendAuthRaven(authToken, 'Easy there', breadcrumb);
};

export const sendAlreadyUpdatedIntegrationError = (authToken, integrationId) => {
  const breadcrumb = {
    message: 'Integration was already updated',
    category: 'Already mutated',
    data: {integrationId}
  };
  return sendAuthRaven(authToken, 'Easy there', breadcrumb);
};

export const sendAlreadyArchivedTeamError = (authToken, teamId) => {
  const breadcrumb = {
    message: 'Team was already archived',
    category: 'Team Already Archived',
    data: {teamId}
  };
  return sendAuthRaven(authToken, 'Easy there', breadcrumb);
}

export const sendAlreadyInactivatedUserError = (authToken, userId) => {
  const breadcrumb = {
    message: 'That user is already inactive. cannot inactivate twice',
    category: 'Already inactive',
    data: {userId}
  };
  return sendAuthRaven(authToken, 'Easy there', breadcrumb);
}
