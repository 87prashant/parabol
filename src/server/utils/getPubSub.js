import getDotenv from '../../universal/utils/dotenv';
import {RedisPubSub} from 'graphql-redis-subscriptions';

getDotenv();

let pubsub;
const getPubSub = () => {
  if (!pubsub) {
    pubsub = new RedisPubSub({
      connection: process.env.REDIS_URL
    });
  }
  return pubsub;
};
export default getPubSub;
