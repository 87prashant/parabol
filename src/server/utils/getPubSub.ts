import getDotenv from '../../universal/utils/dotenv'
import {RedisPubSub} from 'graphql-redis-subscriptions'

getDotenv()

let pubsub: RedisPubSub | undefined
const getPubSub = () => {
  if (!pubsub) {
    pubsub = new RedisPubSub({
      connection: process.env.REDIS_URL
    })
  }
  return pubsub
}
export default getPubSub
