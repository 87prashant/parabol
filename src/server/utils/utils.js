import mime from 'mime-types';
import {APP_MAX_AVATAR_FILE_SIZE} from '../../universal/utils/constants';

// todo put this in a legitify schema
export function validateAvatarUpload(contentType, contentLength) {
  if (typeof process.env.CDN_BASE_URL === 'undefined') {
    throw new Error('CDN_BASE_URL environment variable is not defined');
  }
  if (!contentType || !contentType.startsWith('image/')) {
    throw new Error('file must be an image');
  }
  const ext = mime.extension(contentType);
  if (!ext) {
    throw new Error(`unable to determine extension for ${contentType}`);
  }
  if (contentLength > APP_MAX_AVATAR_FILE_SIZE) {
    throw new Error('avatar image is too large');
  }
  return ext;
}
