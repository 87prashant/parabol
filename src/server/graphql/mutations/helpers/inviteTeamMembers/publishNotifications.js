import getPubSub from 'server/utils/getPubSub';
import {NOTIFICATIONS_ADDED, NOTIFICATIONS_CLEARED} from 'universal/utils/constants';
import mergeObjectsWithArrValues from 'universal/utils/mergeObjectsWithArrValues';

const publishNotifications = (notificationsToSend) => {
  const {
    reactivation,
    notificationsToClear,
    teamInvite,
    approval
  } = notificationsToSend;

  const allNotificationsToAdd = mergeObjectsWithArrValues(reactivation, teamInvite, approval);
  Object.keys(allNotificationsToAdd).forEach((userId) => {
    const notificationsAdded = {
      notifications: allNotificationsToAdd[userId]
    };
    getPubSub().publish(`${NOTIFICATIONS_ADDED}.${userId}`, {notificationsAdded});
  });
  Object.keys(notificationsToClear).forEach((userId) => {
    const notificationsCleared = notificationsToClear[userId];
    getPubSub().publish(`${NOTIFICATIONS_CLEARED}.${userId}`, {notificationsCleared});
  });
};

export default publishNotifications;
