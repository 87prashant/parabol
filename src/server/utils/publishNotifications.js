import getPubSub from 'server/utils/getPubSub';
import {NOTIFICATIONS_ADDED, NOTIFICATIONS_CLEARED} from 'universal/utils/constants';

const publishNotifications = ({notificationsToAdd, notificationsToClear}) => {
  Object.keys(notificationsToAdd).forEach((userId) => {
    const notificationsAdded = {
      notifications: notificationsToAdd[userId]
    };
    getPubSub().publish(`${NOTIFICATIONS_ADDED}.${userId}`, {notificationsAdded});
  });
  Object.keys(notificationsToClear).forEach((userId) => {
    const notificationsCleared = {
      deletedIds: notificationsToClear[userId]
    };
    getPubSub().publish(`${NOTIFICATIONS_CLEARED}.${userId}`, {notificationsCleared});
  });
};

export default publishNotifications;
