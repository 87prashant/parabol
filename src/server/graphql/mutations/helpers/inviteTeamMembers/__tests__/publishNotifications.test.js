import makeMockPubSub from 'server/__mocks__/makeMockPubSub';
import publishNotifications from 'server/graphql/mutations/helpers/inviteTeamMembers/publishNotifications';
import * as getPubSub from 'server/utils/getPubSub';
import {REACTIVATED} from 'universal/utils/constants';

describe('publishNotifications', () => {
  test('publishes new notifications to the message bus', () => {
    // SETUP
    const mockPubSub = makeMockPubSub();
    getPubSub.default = () => mockPubSub;
    const notificationsToAdd = {
      reactivation: {
        1: [{type: REACTIVATED}, {type: REACTIVATED}],
        2: [{type: REACTIVATED}]
      }
    };

    // TEST
    publishNotifications({notificationsToAdd});

    // VERIFY
    expect(mockPubSub.publish.mock.calls).toMatchSnapshot();
  });
  test('publishes cleared notifications to the message bus', () => {
    // SETUP
    const mockPubSub = makeMockPubSub();
    getPubSub.default = () => mockPubSub;
    const notificationsToClear = {
      1: [{deletedId: 1}, {deletedId: 2}],
      2: [{deletedId: 3}]
    };

    // TEST
    publishNotifications({notificationsToClear});

    // VERIFY
    expect(mockPubSub.publish.mock.calls).toMatchSnapshot();
  });
});

