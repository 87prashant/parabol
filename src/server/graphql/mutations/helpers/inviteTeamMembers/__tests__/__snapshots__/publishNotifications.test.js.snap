// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`publishNotifications publishes cleared notifications to the message bus 1`] = `Array []`;

exports[`publishNotifications publishes new notifications to the message bus 1`] = `
Array [
  Array [
    "notificationsAdded.1",
    Object {
      "notificationsAdded": Object {
        "notifications": Array [
          Object {
            "type": "REACTIVATED",
          },
          Object {
            "type": "REACTIVATED",
          },
        ],
      },
    },
  ],
  Array [
    "notificationsAdded.2",
    Object {
      "notificationsAdded": Object {
        "notifications": Array [
          Object {
            "type": "REACTIVATED",
          },
        ],
      },
    },
  ],
]
`;
